### ISSUE TITLE
Narrow navigation types for `beforeNavigate` and `afterNavigate`

### ISSUE THREAD

**Issue by Rich-Harris** [sveltejs/kit#7505]
### Describe the problem

After #6813, `NavigationType` is `'enter' | 'link' | 'goto' | 'popstate' | 'leave'`. But this doesn't quite tell the full story — `afterNavigate` can never be called with a `leave` navigation, and `beforeNavigate` can never be called with an `enter` navigation. 

Similarly, the `$navigating` store value can never have `type === 'enter'`.

It would be nice to guard against this sort of mistake (https://github.com/sveltejs/kit/pull/6813#issuecomment-1303945524 for more of an explanation):

```js
beforeNavigate((navigation) => {
  if (navigation.type === 'enter') {
    // this doesn't make any sense!
  }
});

afterNavigate((navigation) => {
  if (navigation.type === 'leave') {
    // nor does this
  }
});
```

### Describe the proposed solution

Not sure. Generics?

### Importance

nice to have

**Comment by dummdidumm** [comment]
I think the easiest solution would be two different interfaces, one for `beforeNavigate` and `$navigating`, and one for `afterNavigate`. Strictly speaking this would be a breaking change.
